node/core/provisioner/src/disputes/prioritized_selection/mod.rs

```diff
@@ -0,0 +1,445 @@
+			if votes_len + total_votes_len > MAX_DISPUTE_VOTES_FORWARDED_TO_RUNTIME {
+				// we are done - no more votes can be added
+				return result
+			}
+			result.insert((session_index, candidate_hash), selected_votes);
+			total_votes_len += votes_len
+		}
+	}
+
+	result
+}
+
+/// Contains disputes by partitions. Check the field comments for further details.
+#[derive(Default)]
+pub(crate) struct PartitionedDisputes {
+	/// Concluded and inactive disputes which are completely unknown for the Runtime.
+	/// Hopefully this should never happen.
+	/// Will be sent to the Runtime with FIRST priority.
+	pub inactive_unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+	/// Active disputes completely unknown onchain.
+	/// Will be sent to the Runtime with SECOND priority.
+	pub active_unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+	/// Active disputes unconcluded onchain.
+	/// Will be sent to the Runtime with THIRD priority.
+	pub active_unconcluded_onchain: Vec<(SessionIndex, CandidateHash)>,
+	/// Active disputes concluded onchain. New votes are not that important for
+	/// this partition.
+	/// Will be sent to the Runtime with FOURTH priority.
+	pub active_concluded_onchain: Vec<(SessionIndex, CandidateHash)>,
+	/// Inactive disputes which are known onchain. These are not
+	/// interesting and won't be sent to the Runtime.
+	pub inactive_known_onchain: Vec<(SessionIndex, CandidateHash)>,
+}
+
+impl PartitionedDisputes {
+	fn new() -> PartitionedDisputes {
+		Default::default()
+	}
+
+	fn into_iter(self) -> impl Iterator<Item = (SessionIndex, CandidateHash)> {
+		self.inactive_unknown_onchain
+			.into_iter()
+			.chain(self.active_unknown_onchain.into_iter())
+			.chain(self.active_unconcluded_onchain.into_iter())
+			.chain(self.active_concluded_onchain.into_iter())
+		// inactive_known_onchain is dropped on purpose
+	}
+}
+
+fn secs_since_epoch() -> Timestamp {
+	match SystemTime::now().duration_since(UNIX_EPOCH) {
+		Ok(d) => d.as_secs(),
+		Err(e) => {
+			gum::warn!(
+				target: LOG_TARGET,
+				err = ?e,
+				"Error getting system time."
+			);
+			0
+		},
+	}
+}
+
+fn partition_recent_disputes(
+	recent: Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	onchain: &HashMap<(SessionIndex, CandidateHash), DisputeState>,
+) -> PartitionedDisputes {
+	let mut partitioned = PartitionedDisputes::new();
+
+	// Drop any duplicates
+	let unique_recent = recent
+		.into_iter()
+		.map(|(session_index, candidate_hash, dispute_state)| {
+			((session_index, candidate_hash), dispute_state)
+		})
+		.collect::<HashMap<_, _>>();
+
+	// Split ACTIVE from CONCLUDED disputes
+	let (active, inactive): (
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	) = unique_recent
+		.into_iter()
+		.map(|((session_index, candidate_hash), dispute_state)| {
+			(session_index, candidate_hash, dispute_state)
+		})
+		.partition(|(_, _, status)| !dispute_is_inactive(status, &secs_since_epoch()));
```

Comment:
I would feel better if we did not call this function through the iteration, but only once before and cache the result in a variable.

Reply:
I assume you mean `secs_since_epoch()`. That's a very good point. We might drop silently some disputes the way I wrote this.